### ISSUE TITLE
Have Renovate create and use dependency + dependency type labels automatically

### ISSUE THREAD

**Issue by HonkingGoose** [renovatebot/renovate#7485]
**What would you like Renovate to be able to do?**

I want Renovate to figure out labeling its pull requests automatically.

Dependabot just creates/uses labels on my repository as needed, with no input from me.
From the [Dependabot docs on setting custom labels](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/customizing-dependency-updates#setting-custom-labels):

> By default, GitHub Dependabot raises all pull requests with the `dependencies` label. If more than one package manager is defined, Dependabot includes an additional label on each pull request. This indicates which language or ecosystem the pull request will update, for example: `java` for Gradle updates and `submodules` for git submodule updates. GitHub Dependabot creates these default labels automatically, as necessary in your repository.

From the [Renovate docs on labels](https://docs.renovatebot.com/configuration-options/#labels):

> By default, Renovate won't add any labels to its PRs. If you want Renovate to do so then define a `labels` array of one or more label strings. If you want the same label(s) for every PR then you can configure it at the top level of config. However you can also fully override them on a per-package basis.

**Did you already have any implementation ideas?**

Yes, I made a basic Given, When, Then template in Gherkin-like syntax that you can take a look at.
It does not cover all cases, but it should give you a general idea of the kind of functionality that I would like.

https://gist.github.com/HonkingGoose/b8263217a543eeed0ee4ed9d2ddeeeb0

Use the `raw` view to see the Gist with proper formatting.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

I could manually configure the labels, but this is difficult for first time users like me, the Renovate docs are not really clear on how to do this precisely anyways... :cry: 

**Is this a feature you'd be interested in implementing yourself?**

I cannot write the code, however I have made a basic Given, When, Then template for this feature request.

**Comment by rarkins** [comment]
This is definitely not nice behavior to do *by default* (a lot of projects wouldn't want their label list to be polluted by so many new ones), but it sounds fine as an opt-in idea.

One challenge to overcome is what to do when a Pull Request contains more than one language/type - include them all?

**Comment by HonkingGoose** [comment]
> This is definitely not nice behavior to do by _default_ (a lot of projects wouldn't want their label list to be polluted by so many new ones), but it sounds fine as an opt-in idea.

I can understand that, making this _opt-in_ is fine!

> One challenge to overcome is what to do when a Pull Request contains more than one language/type - include them all?

The quickest thing to program would be to include them all, but this leads to a lot of labels...

Maybe something like this is more elegant for the grouped updates:

```
Scenario: a pull request is opened that groups multiple languages/ecosystems into one pull request
  When Renovate opens a pull request
  And the pull request groups updates for multiple ecosystems
  Then Renovate does not apply the ecosystem labels
  But instead Renovate applies the following labels: `dependencies` `grouped_updates`
```

**Comment by rarkins** [comment]
> The quickest thing to program would be to include them all, but this leads to a lot of labels...

I don't think that a lot of labels in a PR is a problem, if you are OK with having lots of labels in your *repository label list*. So for someone who already opts-in to label PRs with languages (which means they're OK with lots of labels in their repo), *and* chooses to configure Renovate to group across languages, then this may be something they just need to live with. For now I vote to merge the label list in such cases.

**Comment by HonkingGoose** [comment]
> if you are OK with having lots of labels in your repository label list.

By the way, the way I envision it, Renovate should only create labels for items that are actually in use in the repository.
No sense in creating labels for Docker, PHP, Nuget, etc if you only use JavaScript in that repository. :+1: 

> For now I vote to merge the label list in such cases.
That's fine by me, if labels is what the user wants, then labels the user shall get. :zany_face:

**Comment by rarkins** [comment]
> By the way, the way I envision it, Renovate should only create labels for items that are actually in use in the repository.
> No sense in creating labels for Docker, PHP, Nuget, etc if you only use JavaScript in that repository. 👍

Yes, that's how it would work. GitHub's API doesn't actually need you to explicitly create labels in a repo - you just assign them to a PR and they get added implicitly if they are new.

**Comment by HonkingGoose** [comment]
I found feature request #6503, which is sort of similar, but that makes the label per dependency type a manual job, instead of Renovate doing it for the user. Just mentioning this, because it seems similar.

**Comment by HonkingGoose** [comment]
Well, I have had a lot of time to think about this feature request that I raised... 😄 

I gained experience with helping to manage a repository that has a lot of labels.
I understand fully why **rarkins** said adding labels for dependencies and ecosystems is not a nice thing to do by default.

Imagine if we got even more labels on the `renovatebot/renovate` repository 😱 👻 .
The label list is already long enough, without having even more labels for `dependency: NodeJS`, `dependency: docker` etc. labels...

I'm also not sure who actually benefits from this feature:

- Small repositories would get labels, but a small repo doesn't really need extensive categorization.
- Medium repositories might benefit from having some way to filter to the relevant dependency updates.
- For large repositories it's just plain unpractical, they would just use the Renovate dashboard anyways to get an overview.

So maybe this feature is only helpful for medium size repositories, and only up-till a certain size.

Maybe we should focus instead of making the Renovate documentation better, to allow a user to configure the labels for certain groups of packages themselves.

**Comment by rarkins** [comment]
If we allow labels to use templates then that would allow users to self configure this type of thing using package rules.

**Comment by HonkingGoose** [comment]
Seeing as we've closed a duplicate issue, I'll copy/paste the iscussion from the close issue here:

A user (@bobvandevijver) opened a new issue, with the following body text:

> As with the created MR/PRs I would like to configure a label (or multiple) for the issues created for the dependency dashboard. I currently run renovate on 14 different projects, so being abled to open them using a group label would be awesome.

@rarkins replied:

> We don't currently have a `dashboard: {}` nested configuration object, so until then I think it would be OK to add `dependencyDashboardLabels` as a new field.
> 
> Something different we may want to do is "ensure" the labels. Normally we only add labels during PR creation, which gives users the flexibility to remove them afterwards without us continuously adding them back. For the dashboard I think it would be better to sync them each time so that if people add labels config _after_ the issue exists then they still get added.

**Comment by rarkins** [comment]
whoops, I misread this - my mistake. I think this is different and will reopen the other one.
